Fo #34914, upgrade to specific core

jfboismenu opened the pull request:
It is now possible to pick the version of core you want to upgrade or downgrade to.

jfboismenu commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -15,36 +15,35 @@
 import sys
```
File still needs to be renamed, but I didn't do it right away because I was afraid it would mess the diff in Github.

jfboismenu commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -53,33 +52,37 @@ def __init__(self):
         """
-        return self._run(log, True)
-    
+        return self._run(log, True, parameters[0] if len(parameters) else None)
```
The non interactive version never cared for the number of arguments, so we won't care if there is more than 1 to make sure we don't break anything.

manneohrstrom replied:
I think we should have an arg based syntax:

```
./tank core --version=v1.2.3
./tank core --version="sgtk:descriptor:dev?path=/Users/manne/Documents/work_dev/toolkit/tk-core"
```

jfboismenu commented:
This is that the diff looks like between
- original upgrade to latest message(left)
- the new update to latest message(center), which is more neutral than UPGRADE since it can also be used to downgrade
- new the update to a specific version message (right)

<img width="1918" alt="screen shot 2016-03-04 at 8 09 14 am" src="https://cloud.githubusercontent.com/assets/8126447/13529438/3d4203f6-e1ea-11e5-88f2-00be616ee928.png">

robblau commented on python/tank/deploy/tank_command.py:
```diff
@@ -47,9 +47,9 @@
+BUILT_IN_ACTIONS = [setup_project.SetupProjectAction,
                     setup_project_wizard.SetupProjectFactoryAction,
-                    core_upgrade.CoreUpgradeAction, 
+                    core_upgrade.CoreUpdateAction,
```
Why the class rename?  I get that it is slightly more accurate, but this is part of our public API so we shouldn't change it without good reason and ideally backward compatibility.  I think there are people out there who deal with things like core updates programmatically and I think this would break their code.

manneohrstrom replied:
I might have missed something and/or things may have changed over time, but heads up that this was originally specifically designed to *not* be part of the public API (C++ pimpl style). When you use the tk API command factory, you get a `SgtkSystemCommand` back that you use to interact with the underlying tk command implementation:

```python
>>> tk
<Sgtk Core v0.16.42@0x102c7e810 Config /mnt/wintermute/configs/sha_is_in_the_heart>
>>> tk.list_commands()
['setup_project', 'setup_project_factory', 'core', 'localize', 'share_core', 'attach_to_core', 'validate', 'cache_apps', 'clear_shotgun_menu_cache', 'app_info', 'install_app', 'install_engine', 'updates', 'configurations', 'synchronize_folders', 'unregister_folders', 'clone_configuration', 'copy_apps', 'cache_yaml', 'get_entity_commands']
>>> tk.get_command("core")
<tank.deploy.tank_command.SgtkSystemCommand object at 0x108521310>
>>> cmd = tk.get_command("core")
>>> dir(cmd)
['_SgtkSystemCommand__internal_action_obj', '_SgtkSystemCommand__log', '__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'category', 'description', 'execute', 'name', 'parameters', 'set_logger']
>>> cmd.execute({})
INFO 
INFO Welcome to the Shotgun Pipeline Toolkit update checker!
INFO This script will check if the Toolkit Core API installed
INFO in /mnt/wintermute/configs/sha_is_in_the_heart
INFO is up to date.

...

INFO 
{'status': 'updated', 'new_version': 'v0.17.8'}

```

jfboismenu replied:
I assumed that the way the Desktop runs commands was the official one, which means that all these contants and classes are not visible from the public API.

Maybe I'm being a bit too pedantic about this, but this is supposed to be internal to core and the terminology wasn't accurate anymore, so I feel like it would be better to fix it.

jfboismenu replied:
I was also very careful not to update the status messages for that very reason, since they are public.

manneohrstrom commented:
General notes on the copy:

Localized configs are now the norm. The copy in the command is from a time when localized configs were only used when you wanted to test a new core or "freeze" a core for a project. I think we should change this. @kporangehat have gotten client feedback about this message being confusing and making people worried that they may break other projects when they upgrade. (it would also be good to get KPs eyes on the copy in general, he is great with words).

My suggestions: Hide the "Please note...." paragraph unless the config for which we are running the upgrade is using a shared core. Maybe we also want to tweak the copy? (I defer to @kporangehat )

There is still a (super edge case?) scenario where a localized config is used by another project, but i don't think we should think of this as a justification to *keep* the "Please note...." paragraph. In this case, there is still the "Please note that this upgrade will affect all projects..." displayed right before the y/n prompt.

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -15,36 +15,35 @@
-        
-class CoreUpgradeAction(Action):
+
+class CoreUpdateAction(Action):
```
this is nitpicky, sorry! But like @robblau I am also wondering if it's worth changing the name. I am reading https://support.apple.com/en-gb/HT201564 and http://english.stackexchange.com/questions/13353/are-there-any-differences-between-update-and-upgrade and to me it feels like Upgrade is actually more appropriate than update. The core command delivers both features and bug fixes, it adds new functionality (e.g. upgrade) as well as making sure a user is kept up to date with the latest fixes (e.g. update).

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -162,88 +174,90 @@ def _run(self, log, suppress_prompts):
     (UP_TO_DATE,                    # all good, no update necessary  
-     UPGRADE_POSSIBLE,              # more recent version exists
-     UPGRADE_BLOCKED_BY_SG          # more recent version exists but SG version is too low.
+     UPDATE_POSSIBLE,              # more recent version exists
```
it seems the change gets a lot smaller if we just keep "Upgrade" :)

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -162,88 +174,90 @@ def _run(self, log, suppress_prompts):
-            
-    def __get_latest_version(self):
+
+    def __get_core_version(self, core_version):
```
maybe this method name should mention app store?

manneohrstrom replied:
i think we should do this change off the `develop/0.18` branch, which supports core descriptors. then we can just the descriptor framework to resolve the code rather than having to hard code all this stuff. We would also get the descriptor based core syntax pretty much for free (e.g we get both `./tank core --version=v1.2.3`
and `./tank core --version="sgtk:descriptor:dev?path=/Users/manne/Documents/work_dev/toolkit/tk-core"`)

manneohrstrom commented:
Hey! I added some comments. Most importantly, I think we should do this change as part of 0.18, because 0.18 supports core descriptors, which will help simplify this code and reduce duplication. I hope it's not a huge reshuffle to move this code across to 0.18 - hopefully in the end it will mean less code and improved reuse! At the very least I think it's worth taking a quick look at what it would mean practically.

jfboismenu commented:
As a general comments for all these comments: I changed the word Upgrade for a very specific reason: we can now use the core command to DOWNGRADE a version of the core. That's why I fixed the terminology.

jfboismenu commented:
Also, the code was already inconsistent in the usage of upgrade vs update both in the code and in the text being printed. Using UPDATE everywhere makes it not consistent but also in line with what the command actually allows you to do.

jfboismenu commented:
This will of course be revisited (or rewritten most likely) in the 0.18 branch, but the reason why we were doing this in the current core instead of 0.18 is that there is a current need from certain clients to be able to downgrade the core, hence doing it right away in 0.17.

manneohrstrom commented:
cool, if there is a reason for it to go in asap, then just ignore me :) this ticket was part of the whole plugin epic and that got me confused! In that case, we should probably have a ticket to revisit and refactor this in 0.18.

jfboismenu commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -15,36 +15,46 @@
-        
-class CoreUpgradeAction(Action):
+
+# FIXME: This should be refactored into something that can be used by other commands.
```
Indeed.

jfboismenu commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -53,76 +66,108 @@ def __init__(self):
+        parser = TkOptParse()
+        parser.set_usage(optparse.SUPPRESS_USAGE)
+        parser.add_option("-v", "--version", type="string", default=None)
+        options, args = parser.parse_args()
```
Here's what the following inputs will yield (portions omitted for brievity)

```
> ./tank core -v

Welcome to the Shotgun Pipeline Toolkit!
For documentation, see https://support.shotgunsoftware.com
Starting toolkit for path '/Users/jfboismenu/shotgun/jf/bbb/pipeline'
- Running as user 'jfboismenu'
- Using configuration 'Primary' and Core v0.7.8
- Setting the Context to Big Buck Bunny.
- Running command core...


----------------------------------------------------------------------
Command: Core
----------------------------------------------------------------------


ERROR: -v option requires an argument

> ./tank core -v1
[...]
ERROR: version string should always start with 'v'

> ./tank core -vv2
[...]
ERROR: Could not find version 'v2' of the Core API in the app store!

> ./tank core -version=1
[...]
ERROR: version string should always start with 'v'

> ./tank core --version=1
[...]
ERROR: version string should always start with 'v'

> ./tank core --version=v1
[...]
ERROR: Could not find version 'v1' of the Core API in the app store!

> ./tank core --version=v1.2.3 --not-an-option
[...]
ERROR: no such option: --not-an-option
```

manneohrstrom replied:
you previously mentioned backwards compatibility - previously someone running `./tank core --foo --bar` would be fine, after the upgrade they will get an error. Something we should be worried about?

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -15,36 +15,49 @@
+
+# FIXME: This should be refactored into something that can be used by other commands.
+class TkOptParse(optparse.OptionParser):
+    def __init__(self, *args, **kwargs):
```
This is great to see! Nince one! Can you add some docstrings? :) Also, I think if we should do it we should do it properly; i suggest moving it to action_base and maybe add some helpers to the base class to make it easy to use?

jfboismenu replied:
Before turning this into an API feature, maybe we should look at what it means using OptionGroups and OptionParser in a more general sense in Toolkit first.

Docstrings done.

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -53,33 +66,54 @@ def __init__(self):
+
+        :returns: The core version. None if --version wasn't specified.
+        """
+        parser = TkOptParse()
```
similar to how we define `self.parameters` today and then run `validate_parameters()` - https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/deploy/tank_commands/action_base.py#L108 

Could we do something analogous with the arg flags?

jfboismenu replied:
As I said, I'd rather wait to have a better idea of the direction we'd want to take with OptionParser in general with Toolkit.

manneohrstrom replied:
Can we be agile? I don't think it needs to be super complicated.

We have several commands that take *optional* flags. Quick grep gave me `--version=xxx`, `--preserve-yaml`, `--backout`, `--full`, `--symlink`, `--force`, `--no-storage-check`, `--all`, `--filter=xxx`, `--external=xxxx`. Just supporting that would be a really great useful first step.

This is always a pre-flight thing, we look for certain flags, they are always optional, we parse, remove args we discover from the from the args string and then return it back to the main tank command for further processing.

Sorry if i am missing something!

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -88,41 +122,52 @@ def _run(self, log, suppress_prompts):
-        installer = TankCoreUpgrader(code_install_root, log)
+
+        if not pipelineconfig_utils.is_localized(self.tk.pipeline_configuration.get_path()):
+            log.info("Please note that when you update the core API, you typically affect "
```
:+1: :smile:

manneohrstrom commented on python/tank/deploy/tank_commands/core_upgrade.py:
```diff
@@ -162,88 +207,96 @@ def _run(self, log, suppress_prompts):
+
+        if version_found is None:
+            if version_requested:
+                raise TankError("Could not find version '%s' of the Core API in the app store!" % version_requested)
```
break this line?

manneohrstrom commented:
cr looks great - i think we should move the opt parse stuff into base and add a helper method there while we are at it. Also i think if we could get @kporangehat to take a quick look at the final output i think that would be great. Now that we are making changes in here anyways :)
